### ISSUE TITLE
dkg.js & Nextjs13 Server Side Rendering conflicts?

### ISSUE THREAD

**Issue by FamousAmos-ChillZone** [OriginTrail/dkg.js#79]
Hello Team,

I am currently endeavoring to integrate the dkg.js module with a Next.js 13 application, yet I am encountering compilation challenges. Given that Next.js operates within a Node.js runtime, I had anticipated a more seamless process.

When attempting to employ the SDK, I am presented with the following error:

error node_modules/assertion-tools/keccak256.js (3:17) @ keccak256
- error Error [ReferenceError]: keccak256 is not defined
    at eval (webpack-internal:///(sc_server)/./node_modules/assertion-tools/keccak256.js:3:28)

As a temporary solution, I have incorporated an Express.js server within the Next.js application. However, this custom server architecture does restrict some of the novel benefits provided by the Next.js app router setup. Moreover, it generates issues unless the automatic page routing is deactivated.

I am using the standard setup for API routes within NextJS13. 

I am eager to utilize the DKG SDK within the standard routing environment of Next.js, and any guidance on resolving this issue would be greatly appreciated.

Best Regards,
Famos Amos

**Comment by branarakic** [comment]
Thanks for reporting this @FamousAmos-ChillZone , we'll investigate asap on our end

**Comment by brkagithub** [comment]
Hey @FamousAmos-ChillZone, I am currently in the process of setting up a project with NextJS 13 and will try to set up the DKG SDK with it and see if it works (will edit this comment or add another one when I'm done). Can you provide some more information: the NodeJS version you used, and some more details about how you set up your Next application. It's the first time for me using the newest NextJS 13, but from what I know, you can still use the older NextJS api routes, but you can also use React server components to call the backend - which one of these did you use?

**Comment by FamousAmos-ChillZone** [comment]
Hello!

My apologies for the late response. I'm using node V18.12.1 but i've encountered the issue on different versions such as:

node v 6.0.0
node v 18.12.1
node v 19.3.0

I do see that you're encountering the same issues as I am thus it appears I have successful found a bug. I am excited to hear your response. In the meanwhile I will be using the custom express server to continue building out my application.

Thanks,
Famous Amos

**Comment by brkagithub** [comment]
Hey @FamousAmos-ChillZone,

Thanks for the response. I have located the bug and managed to fix it on my local setup, we just need a day/few days to review the code that we're going to be changing in dkg.js and the fix should be live, I'll notify you then and close the issue😆 

Thank you very much for reporting, and please continue to report any issues you run into, especially with NextJS 13 as we definitely want to support developing in modern web frameworks like NextJS 13.

**Comment by NZT48** [comment]
Hey @FamousAmos-ChillZone,

We have released new DKG.js version 6.0.6. The fix for this issue is included in the release, so can you please bump the version of dkg.js and try to see if it is resolved on your side too. 

Thank you for the issue report and all the support :)

#TraceOn

**Comment by FamousAmos-ChillZone** [comment]
Hey team, sorry for the late response.

Awesome!! I upgraded the version and attempted to compile once more. It does indeed complie using the native standard setup so this is an AWESOME step forward. I do believe there is a bit more compatibility issues to mention that seem to not bring functionality. Here is the error I am seeing directly after embedding the DKG client into my Nextjs13 application and pulling in the standard "node information" request:

------------------------------------ERROR START------------------------------------
- wait compiling...
- warn ./node_modules/cross-fetch/node_modules/node-fetch/lib/index.js
Module not found: Can't resolve 'encoding' in '/***/***/***/Studies/React/dkgnext/node_modules/cross-fetch/node_modules/node-fetch/lib'

Import trace for requested module:
./node_modules/cross-fetch/node_modules/node-fetch/lib/index.js
./node_modules/cross-fetch/dist/node-ponyfill.js
./node_modules/cross-fetch/dist/node-polyfill.js
./node_modules/web3-providers-http/lib/index.js
./node_modules/web3-core-requestmanager/lib/index.js
./node_modules/web3-core/lib/index.js
./node_modules/web3/lib/index.js
./node_modules/dkg.js/services/blockchain-service/implementations/browser-blockchain-service.js
./node_modules/dkg.js/services/blockchain-service/blockchain-interface.js
./node_modules/dkg.js/services/base-service-manager.js
./node_modules/dkg.js/index.js
./components/Call.js
./app/page.js

./node_modules/keyv/src/index.js
Critical dependency: the request of a dependency is an expression

Import trace for requested module:
./node_modules/keyv/src/index.js
./node_modules/cacheable-request/src/index.js
./node_modules/swarm-js/node_modules/got/dist/source/core/index.js
./node_modules/swarm-js/node_modules/got/dist/source/create.js
./node_modules/swarm-js/node_modules/got/dist/source/index.js
./node_modules/swarm-js/lib/files.js
./node_modules/swarm-js/lib/api-node.js
./node_modules/web3-bzz/lib/index.js
./node_modules/web3/lib/index.js
./node_modules/dkg.js/services/blockchain-service/implementations/browser-blockchain-service.js
./node_modules/dkg.js/services/blockchain-service/blockchain-interface.js
./node_modules/dkg.js/services/base-service-manager.js
./node_modules/dkg.js/index.js
./components/Call.js
./app/page.js

./node_modules/rdf-canonize/lib/index.js
Module not found: Can't resolve 'rdf-canonize-native' in '/***/***/***/React/dkgnext/node_modules/rdf-canonize/lib'

Import trace for requested module:
./node_modules/rdf-canonize/lib/index.js
./node_modules/rdf-canonize/index.js
./node_modules/jsonld/lib/jsonld.js
./node_modules/jsonld/lib/index.js
./node_modules/dkg.js/services/utilities.js
./node_modules/dkg.js/managers/asset-operations-manager.js
./node_modules/dkg.js/index.js
./components/Call.js
./app/page.js
(node:45521) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `node --trace-deprecation ...` to show where the warning was created)

------------------------------------ERROR END------------------------------------


Here is the code for the example application. The name of the component is **Call**:

------------------------------------DKG SDK As A Component Start------------------------------------

const DKG = require('dkg.js');
const dkg = new DKG({
  endpoint: 'http://***', // gateway node URI
  port: 8900,
  blockchain: {
    name: 'otp::testnet', // or otp::mainnet, otp::testnet, otp::devnet
    publicKey: '***', // not required in browser, metamask used instead
    privateKey:
      '***', // not required in browser, metamask used instead
  },
});

export default async function call() {
  const nodeInfo = await dkg.node.info();
  return JSON.stringify(nodeInfo);
}
------------------------------------DKG SDK As A Component END------------------------------------

Here is where the component is called:

------------------------------------Root Page JS File Start------------------------------------

import Image from 'next/image';
import Link from 'next/Link';
import Call from '../components/Call';

export default function Home() {
  return (
    <>
      <Link href="/users">Users</Link>
      <Link href="/dashboard">Dashboard</Link>
      <Link href="/calculator">Calculator</Link>
      <Call />
    </>
  );
}
------------------------------------Root Page JS File END ------------------------------------

Thus far, since the changes you beautiful folks have made the SDK at-least doesn't crash the application anymore. Seems like it just needs some final fine tuning and we're all good to go?

**Comment by brkagithub** [comment]
hey @FamousAmos-ChillZone, 

While fixing the bug and testing if it works I also noticed the (almost) same warnings. Since it wasn't actually impacting the functionality, we decided to postpone fixing it, but be assured that it's on our radar and idk if I can give you an exact estimate, but I believe we will take a closer look at it in the upcoming months (or even this month). Personally I like NextJS a lot, never got to try out the version 13 until looking at this bug, so I might even try to create something with the SDK and NextJS 13 in my free time, and try to locate the issue meanwhile 😸

Thanks a lot for reporting the issue in such great detail, helps a lot to reproduce it and fix it, and if you encounter any more issues feel free to write here or ping us on Discord!
